tools/tests/py2fgen/test_dycore_wrapper.py

```diff
@@ -1602,16 +1604,16 @@ def test_granule_solve_nonhydro_multi_step_regional(
             divdamp_fac_o2=initial_divdamp_fac,
             ndyn_substeps=ndyn_substeps,
             idyn_timestep=i_substep,
-            nnow=nnow,
-            nnew=nnew,
         )
         linit = False
         recompute = False
         clean_mflx = False
-        if not is_last_substep:
-            ntemp = nnow
-            nnow = nnew
-            nnew = ntemp
+
+        prognostic_states.swap()
```

Comment:
The wrapper function `dycore_wrapper.solve_nh_run` is the interface to Fortran. 
So the wrapper function needs all information such that it can determine internally wether the `Pair` has to be constructed  as `Pair(ddt_w_adv_ntl1, ddt_w_adv_ntl2)` or `Pair(ddt_w_adv_ntl2, ddt_w_adv_ntl1) `. I think that is possible with `linit` `ndyn_substep` and `idyn_substep` . 

Otherwise we would need to add the additional state that you are creating here outside of it (the Pairs for `ddt_vn_apc` and `ddt_w_adv` as a persistent state from one call to the next on the python side. (The wrapper does this with the grid, but I would try to avoid it for these advection tendencies.

Anyway it is good that is runs, that is good point to start with.

Reply:
Thanks for the explanation. 
Because `solve_nonhydro.time_step` swaps the velocity tendencies internally, it needs to remember which `ddt_w_adv_ntl1` or `ddt_w_adv_ntl2` is the current or next. I agree that it is better not to add additional state just for advection tendencies.
Unless advection tendencies are also swapped outside of `time_step` like `PrognosticState`. But it will feel strange to me that we swap them outside since they are 'local' variables computed only by `solve_nonhydro`.
Possibly the best solution is keeping track of the 0, 1 indices for current and next.

Reply:
I am still a bit worried about the eventual calls from FORTRAN. The wrapper interface takes a list of fields and internally constructs the pairs:
```
def solve_nh_run(...w_now, w_new, ddt_w_adv_ntl1, ddt_w_adv_ntl2)
       ...
       prog_pair = TimestepPar(w_now, w_new) 
       ddt_w_adv_pc = PredictorCorrectorPar(ddt_w_adv_ntl1, ddt_w_adv_ntl2)
       ...
```
where I use only `w` as replacement for the prognostic state.
Because the prognostic state the swap is done externally, every other call the array that is passed as `w_now` will be different, only every second call it will be the same. But for the advection tendencies? Aren't we missing some kind of logic in the _python wrapper_ not in the (real icon4py model code (greenline) and not in the python test, because that will not be executed in Fortran.

Reply:
`solve_nh_run` will be pretty straightforward before this PR is merged because `ntl1` and `ntl2` in `ddt_w_adv_pc` are always set equal to `nnow` and `nnew`. 
To do it correctly when calling from Fortran, instead of doing what I saw in this PR https://github.com/C2SM/icon-exclaim/pull/282,
```
   ! Set time levels of ddt_adv fields for call to velocity_tendencies
      IF (itime_scheme >= 4) THEN ! Velocity advection averaging nnow and nnew tendencies
        ntl1 = nnow
        ntl2 = nnew
      ELSE                        ! Velocity advection is taken at nnew only
        ntl1 = 1
        ntl2 = 1
      ENDIF

      CALL solve_nh_run( 
                         w_now             = p_nh%prog(nnow)%w(:,:,1), &
                         w_new             = p_nh%prog(nnew)%w(:,:,1), &
                         ddt_vn_apc_ntl1   = p_nh%diag%ddt_vn_apc_pc(:,:,1,ntl1), &
                         ddt_vn_apc_ntl2   = p_nh%diag%ddt_vn_apc_pc(:,:,1,ntl2), &
                         ddt_w_adv_ntl1    = p_nh%diag%ddt_w_adv_pc(:,:,1,ntl1), &
                         ddt_w_adv_ntl2    = p_nh%diag%ddt_w_adv_pc(:,:,1,ntl2), &
```
we just need to pass in as below because `ntl1` and `ntl2` are set internally in `SolveNonhydro.time_step`.
```
   ! Set time levels of ddt_adv fields for call to velocity_tendencies
      IF (itime_scheme >= 4) THEN ! Velocity advection averaging nnow and nnew tendencies
        ntl1 = nnow
        ntl2 = nnew
      ELSE                        ! Velocity advection is taken at nnew only
        ntl1 = 1
        ntl2 = 1
      ENDIF

      CALL solve_nh_run( 
                         w_now             = p_nh%prog(nnow)%w(:,:,1), &
                         w_new             = p_nh%prog(nnew)%w(:,:,1), &
                         ddt_vn_apc_ntl1   = p_nh%diag%ddt_vn_apc_pc(:,:,1,1), &
                         ddt_vn_apc_ntl2   = p_nh%diag%ddt_vn_apc_pc(:,:,1,2), &
                         ddt_w_adv_ntl1    = p_nh%diag%ddt_w_adv_pc(:,:,1,1), &
                         ddt_w_adv_ntl2    = p_nh%diag%ddt_w_adv_pc(:,:,1,2), &
```

After we merge this branch using `PredictorCorrectorPair` class, we kind of decouple `ntl1` and `ntl2` from `nnow` and `nnew` and swap when we really need to (I like this because it reflects the time stepping scheme correctly). We now need to know when `ddt_vn_apc_pc` and `ddt_w_adv_pc` are swapped as done in this test
```
        if not (at_initial_timestep and (idyn_timestep == 1)) then
            ddt_w_adv.swap()
        if not (i_substep - 1 == 0):
            ddt_vn_apc.swap()
```
@samkellerhals should also notice this issue.

Reply:
@OngChia I think what you showed in your last example would make sense as the integration code in Fortran.